components/constellation/constellation.rs

```diff
@@ -2996,22 +3013,44 @@ where
         &mut self,
         top_level_browsing_context_id: TopLevelBrowsingContextId,
     ) {
+        debug!("{}: Closing", top_level_browsing_context_id);
         let browsing_context_id = BrowsingContextId::from(top_level_browsing_context_id);
-        self.close_browsing_context(browsing_context_id, ExitPipelineMode::Normal);
-        self.browsers.remove(&top_level_browsing_context_id);
-        if self.active_browser_id == Some(top_level_browsing_context_id) {
-            self.active_browser_id = None;
+        let browsing_context =
+            self.close_browsing_context(browsing_context_id, ExitPipelineMode::Normal);
+        if self.browsers.focused_browser().map(|(id, _)| id) == Some(top_level_browsing_context_id)
+        {
+            self.embedder_proxy
+                .send((None, EmbedderMsg::BrowserUnfocused));
         }
-        let browsing_context = match self.browsing_contexts.get(&browsing_context_id) {
-            Some(bc) => bc,
-            None => {
-                warn!("Browsing context closed before it started");
-                return;
-            },
-        };
-        // https://html.spec.whatwg.org/multipage/#bcg-remove
-        self.browsing_context_group_set
-            .remove(&browsing_context.bc_group_id);
+        self.browsers.remove(top_level_browsing_context_id);
+        // TODO Send the compositor a RemoveBrowser event.
+        self.embedder_proxy.send((
+            Some(top_level_browsing_context_id),
+            EmbedderMsg::BrowserClosed(top_level_browsing_context_id),
+        ));
+
+        if let Some(browsing_context) = browsing_context {
+            // https://html.spec.whatwg.org/multipage/#bcg-remove
+            let bc_group_id = browsing_context.bc_group_id;
+            if let Some(bc_group) = self.browsing_context_group_set.get_mut(&bc_group_id) {
```

Comment:
Ditto. This will also make it easier to see where warnings are needed to be printed.

Reply:
done in 3c910aa12c10955e72021e4ba3bb276c0cc84f59